sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java

```diff
@@ -167,18 +173,83 @@ private TransportCollectProfileOperation getRemoteCollectOperation(DependencyCar
 
         mapBuilder.put(Phase.Execute.name(), executionTimingsMap.immutableMap());
         return new Row1(mapBuilder.immutableMap());
     }
 
+    private static Map<String, Object> extractPhasesTimingsFrom(Map<String, Map<String, Object>> nodesTimings,
+                                                                NodeOperationTree operationTree) {
+        Map<String, Object> allPhases = new TreeMap<>();
+        for (NodeOperation operation : operationTree.nodeOperations()) {
+            ExecutionPhase phase = operation.executionPhase();
+            String phaseName = getPhaseName(phase);
+            Map<String, Object> phaseTimingsAcrossNodes = getPhaseTimingsAcrossNodes(phaseName, nodesTimings);
+
+            if (!phaseTimingsAcrossNodes.isEmpty()) {
+                allPhases.put(phaseName, MapBuilder.newMapBuilder().put("nodes", phaseTimingsAcrossNodes).immutableMap());
+            }
+        }
+
+        ExecutionPhase leafExecutionPhase = operationTree.leaf();
+        String phaseName = getPhaseName(leafExecutionPhase);
+        Map<String, Object> phaseTimingsAcrossNodes = getPhaseTimingsAcrossNodes(phaseName, nodesTimings);
+
+        if (!phaseTimingsAcrossNodes.isEmpty()) {
+            allPhases.put(phaseName, MapBuilder.newMapBuilder().put("nodes", phaseTimingsAcrossNodes).immutableMap());
+        }
+
+        return allPhases;
+    }
+
+    private static String getPhaseName(ExecutionPhase phase) {
+        return String.format(Locale.ROOT, "%d-%s", phase.phaseId(), phase.name());
+    }
+
+    private static Map<String, Object> getPhaseTimingsAcrossNodes(String phaseName,
+                                                                  Map<String, Map<String, Object>> nodesTimings) {
+        MapBuilder<String, Object> timingsForPhaseAcrossNodes = MapBuilder.newMapBuilder();
+        for (String operationNode : nodesTimings.keySet()) {
+            Map<String, Object> timingsForNode = nodesTimings.get(operationNode);
+            if (timingsForNode != null) {
+                Object phaseTiming = timingsForNode.get(phaseName);
+                if (phaseTiming != null) {
+                    timingsForPhaseAcrossNodes.put(operationNode, phaseTiming);
+                }
+            }
+        }
+
+        return timingsForPhaseAcrossNodes.immutableMap();
+    }
+
+    private static Map<String, Map<String, Object>> getNodeTimingsWithoutPhases(Set<String> phasesNames,
+                                                                                Map<String, Map<String, Object>> nodeTimings) {
+        Map<String, Map<String, Object>> nodeTimingsWithoutPhases = new HashMap<>(nodeTimings.size());
+        for (Map.Entry<String, Map<String, Object>> nodeToTimingsEntry : nodeTimings.entrySet()) {
+            nodeTimingsWithoutPhases.put(nodeToTimingsEntry.getKey(), new HashMap<>(nodeToTimingsEntry.getValue()));
+        }
+
+        for (Map<String, Object> timings : nodeTimingsWithoutPhases.values()) {
```

Comment:
I'm wondering if it would be possible to filter the phases already on creation of the timings (L241)

Reply:
Hmm, it's more of a recreation of the map rather than filtering so even if we cram it at the same location somehow it won't be more efficient. 
I think having it separately is more easy to reason about what's going on (and obvious from the method signature we're making a copy of the nodeTimings map)